man/man3/slst_diff.3
.\"
.\" slst_diff.3
.\"
.\" Manpage for slst_diff of Undefined-C library
.\"
.\" By: Juillard Jean-Baptiste (jbjuillard@gmail.com)
.\"
.\" Created: 2017/02/08 by Juillard Jean-Baptiste
.\" Updated: 2018/03/12 by Juillard Jean-Baptiste
.\"
.\" This file is a part free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 3, or
.\" (at your option) any later version.
.\"
.\" There is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file LICENSE.  If not, write to
.\" the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301, USA.
.\"

.TH SLST_DIFF 3 "02/08/17" "Version 0.0" "Manuel du programmeur Undefined-C"

.SH NAME
.B slst_diff, slst_rdiff
.RB "	- Compare exhaustivement deux listes de type " slst_t (3).

.SH SYNOPSIS
.B #include <stdlst.h>

.BI "size_t	slst_diff(slst_t " **lst1 ", mslst_t " **diff1 ", slst_t " **lst2 ,
.BI "mslst_t " **diff2 ", int (" *fcmp ")(const void *, const size_t,"
.B const void *, const size_t));
.br
.BI "size_t	slst_rdiff(slst_t " **lst1 ", mslst_t " **diff1 ", slst_t " **lst2 ,
.BI "mslst_t " **diff2 ", int (" *fcmp ")(const void *, const size_t,"
.B const void *, const size_t));

.SH DESCRIPTION
.B slst_diff
.RI "compare exhaustivement " lst1 " et " lst2 " en fonction de " fcmp " et"
.RI "sauvegarde la liste des differences dans " diff1 " et " diff2 .
.IR lst1 ", " diff1 ", " lst2 ", " diff2 " et " fcmp " doivent etres des"
.RB "arguments valides tels qu'ils sont decrit dans la section " NOTES .
.RB "Dans tous les cas, " errno " est modifier par " slst_diff
.RB "tel que decrit dans la section " ERREURS .

.B slst_rdiff
.RI "compare exhaustivement et recursivement " lst1 " et " lst2 " en fonction"
.RI "de " fcmp " et sauvegarde la liste des differences dans " diff1 " et "
.IR  diff2 .
.IR lst1 ", " diff1 ", " lst2 ", " diff2 " et " fcmp " doivent etres des"
.RB "arguments valides tels qu'ils sont decrit dans la section " NOTES .
.RB "Dans tous les cas, " errno " est modifier par " slst_rdiff
.RB "tel que decrit dans la section " ERREURS .

.SH VALEUR DE RETOUR
.B slst_diff
.RI "retourne le nombre de differences trouve entre " lst1 " et " lst2 .
.RB "Si une erreur se produit , " slst_diff " retourne le nombre de differences"
.R trouve au moment ou se produit l'erreurs. Si aucune difference n'est trouve,
.BR slst_diff "retourne 0."

.BR slst_rdiff
.RI "retourne le nombre de differences trouve entre " lst1 " et " lst2 .
.RB "Si une erreur se produit , " slst_rdiff " retourne le nombre de differences"
.R trouve au moment ou se produit l'erreurs. Si aucune difference n'est trouve,
.BR slst_rdiff "retourne 0."

.SH ERREURS
.TP
.B 0
.R "Aucune erreur ne s'est prouite."
.TP
.B EINVAL
.R "La fonction a ete appele avec un (ou des) parametre(s) invalide(s)."
.TP
.B EOVERFLOW
.RB "Le nombre de difference depasse les limites du type " size_t .

.SH NOTES
.RI "- Pour toutes les fonctions, " lst1 " et " lst2 " doivent etres different"
.RI "de ((slst_t **)(NULL)), notez que " *lstX "=((slst_t *)(NULL))"
.R est valide et designe une liste vide.

.RI "- Pour toutes les fonctions, " diff1 " et " diff2 " doivent etres different"
.RI "de ((slst_t **)(NULL)) et " *diffX " doit etre egale a ((slst_t *)(NULL))."

.R "- Pour toutes les fonctions,"
.IR fcmp " est un pointeur vers une fonction de comparaison."
.RB "Celle-ci prend en parametre deux pointeurs " void " qui adressent les"
.RB "valeurs cles et deux " size_t " qui donnent les tailles des tampons"
.R contenant les valeurs cles.
.RB "Le premier pointeur sur " void " et le premier " size_t " sont ceux de"
.RI "la liste " lst1 " et les seconds ceux de la liste " lst2 .
.IR fcmp " doit retourne un entier positif si la valeur de la liste " lst1
.RI "est supperieur a la valeur de la liste " lst2 ", un entier negatif si la"
.RI "valeur de la liste " lst1 " est inferieur a la valeur de la liste " lst2
.R et 0 si les deux valeurs sont egales. Notez que les deux valeurs n'ont pas
.RI "besoin d'etres strictement egales pour que " fcmp " considere que se soit"
.R le cas.

.SH EXEMPLES

.SH AUTEUR
Juillard Jean-Baptiste

.SH VOIR AUSSI
.BR stdlst "(3), " slst_t "(3), " slst_cmp (3)


src/stdlst/slst_diff.c
/*
** slst_diff function for Undefined-C library
**
** Created: 12/28/2016 by Juillard Jean-Baptiste
** Updated: 2018/03/12 by Juillard Jean-Baptiste
**
** This file is a part free software; you can redistribute it and/or
** modify it under the terms of the GNU General Public License as
** published by the Free Software Foundation; either version 3, or
** (at your option) any later version.
** 
** There is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
** General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; see the file LICENSE.  If not, write to
** the Free Software Foundation, Inc., 51 Franklin Street, Fifth
** Floor, Boston, MA 02110-1301, USA.
*/

#include <libuc/stddef.h>
#include <libuc/errno.h>
#include <stdlib.h>			/* Dev: a modifier après implantation de malloc et de free */
#include <libuc/stdint.h>
#include <libuc/stdlst.h>

size_t	slst_diff(slst_t **lst1,
					mslst_t **diff1,
					slst_t **lst2,
					mslst_t **diff2,
					int (*fcmp)(const void *, const size_t,
								const void *, const size_t))
{
	register slst_t		*tmp1;
	register slst_t		*tmp2;
	register mslst_t	**d1;
	register mslst_t	**d2;
	register size_t		ret;
	register int		err;

	if (!lst1 || !diff1 || *diff1 || !lst2 || !diff2 || *diff2 || !fcmp)
	{
		errno = EINVAL;
		return (0);
	}
	errno = 0;
	tmp1 = *lst1;
	tmp2 = *lst2;
	d1 = diff1;
	d2 = diff2;
	ret = 0;
	while ((tmp1 || tmp2) && !errno)
	{
		if ((tmp1 && tmp2
				&& (*fcmp)(	(const void *)(tmp1->key),
							(const size_t)(tmp1->size),
							(const void *)(tmp2->key),
							(const size_t)(tmp2->size))
				&& !errno)
			|| (!tmp1 && tmp2)
			|| (tmp1 && !tmp2))
		{
			if ((*diff1 = (mslst_t *)malloc(sizeof(mslst_t)))
					== (mslst_t *)(NULL)
				|| (*diff2 = (mslst_t *)malloc(sizeof(mslst_t)))
					== (mslst_t *)(NULL))
			{
				if (!errno)
					errno = ENOMEM;
				err = errno;
				if (*diff1)
					free((void *)(diff1));
				mslst_purge(d1);
				mslst_purge(d2);
				errno = err;
				return (0);
			}
			(*diff1)->kptr = ((tmp1) ? &(tmp1->key) : (void **)(NULL));
			(*diff1)->sptr = ((tmp1) ? &(tmp1->size) : (size_t *)(NULL));
			(*diff1)->next = (mslst_t *)(NULL);
			diff1 = &((*diff1)->next);
			(*diff2)->kptr = ((tmp2) ? &(tmp2->key) : (void **)(NULL));
			(*diff2)->sptr = ((tmp2) ? &(tmp2->size) : (size_t *)(NULL));
			(*diff2)->next = (mslst_t *)(NULL);
			diff2 = &((*diff2)->next);
			if (ret == SIZE_MAX)
			{
			    err = EOVERFLOW;
				mslst_purge(d1);
				mslst_purge(d2);
				errno = err;
				return (0);
			}
			ret++;
		}
		if (tmp1)
			tmp1 = tmp1->next;
		if (tmp2)
			tmp2 = tmp2->next;
		
	}
	return (ret);
}
